lucene/highlighter/src/java/org/apache/lucene/search/uhighlight/CompositePostingsEnum.java

```diff
@@ -0,0 +1,165 @@
+
+
+    private final PostingsEnum postingsEnum;
+    private final int freq;
+    private int position;
+    private int nextPosition;
+    private int positionInc = 1;
+
+    private int startOffset;
+    private int endOffset;
+
+    BoundsCheckingPostingsEnum(PostingsEnum postingsEnum) throws IOException {
+      this.postingsEnum = postingsEnum;
+      this.freq = postingsEnum.freq();
+      nextPosition = postingsEnum.nextPosition();
+      position = nextPosition;
+      startOffset = postingsEnum.startOffset();
+      endOffset = postingsEnum.endOffset();
+    }
+
+    private boolean hasMorePositions() throws IOException {
+      return positionInc < freq;
+    }
+
+    /**
+     * Returns the next position of the underlying postings enum unless
+     * it cannot iterate further and returns NO_MORE_POSITIONS;
+     * @return
+     * @throws IOException
+     */
+    private int nextPosition() throws IOException {
+      position = nextPosition;
+      startOffset = postingsEnum.startOffset();
+      endOffset = postingsEnum.endOffset();
+      if (hasMorePositions()) {
+        positionInc++;
+        nextPosition = postingsEnum.nextPosition();
+      } else {
+        nextPosition = NO_MORE_POSITIONS;
+      }
+      return position;
+    }
+
+  }
+
+  CompositePostingsEnum(BytesRef term, List<PostingsEnum> postingsEnums) throws IOException {
+    this.term = term;
+    queue = new PriorityQueue<BoundsCheckingPostingsEnum>(postingsEnums.size()) {
+      @Override
+      protected boolean lessThan(BoundsCheckingPostingsEnum a, BoundsCheckingPostingsEnum b) {
+        return a.position < b.position;
+      }
+    };
+
+    int freqAdd = 0;
+    for (PostingsEnum postingsEnum : postingsEnums) {
+      queue.add(new BoundsCheckingPostingsEnum(postingsEnum));
+      freqAdd += postingsEnum.freq();
+    }
+    freq = freqAdd;
+  }
+
+  @Override
+  public int freq() throws IOException {
+    return freq;
+  }
+
+  @Override
+  public int nextPosition() throws IOException {
+    int position = NO_MORE_POSITIONS;
+    while (queue.size() >= 1) {
+      queue.top().nextPosition();
+      queue.updateTop(); //the new position may be behind another postingsEnum in the queue
+      position = queue.top().position;
+
+      if (position == NO_MORE_POSITIONS) {
+        queue.pop(); //this postingsEnum is consumed, let's get rid of it
+      } else {
+        break; //we got a new position
+      }
+
+    }
+    return position;
+  }
+
+  @Override
+  public int startOffset() throws IOException {
+    return queue.top().startOffset;
+  }
+
+  @Override
+  public int endOffset() throws IOException {
+    return queue.top().endOffset;
+  }
+
+  @Override
+  public BytesRef getPayload() throws IOException {
+    //The UnifiedHighlighter depends on the payload for a wildcard
+    //being the term representing it
+    return term;
+  }
+
+  @Override
+  public int docID() {
+    return queue.top().postingsEnum.docID();
```

Comment:
I think this one should be held in a field of CompositePostingsEnum initialized in constructor?  Not a big deal though, since I don't think the highlighter is going to call this, let alone sometime after positions are exhausted (which would trigger an NPE in this case).